File: 01-arrays_&_hashing/00238-Product_of_Array_Except_Self/00238-product_of_array_except_self.go
// 238 : Product of Array Except Self
// https://leetcode.com/problems/product-of-array-except-self/

package main

import "fmt"

// SOLUTION
func productExceptSelf(nums []int) []int {
	result := make([]int, len(nums))
	for i := range result {result[i] = 1}

	for i, suf, pre, n:=0, 1, 1, len(nums); i<n; i++ {
		result[i] *= pre;
		pre *= nums[i];
		result[n-1-i] *= suf;
		suf *= nums[n-1-i];
	}

	return result
}


func main() {
	// INPUT
	nums := []int{1,2,3,4}

	// OUTPUT
	result := productExceptSelf(nums)
	fmt.Println(result)
}

File: 01-arrays_&_hashing/00238-Product_of_Array_Except_Self/00238-product_of_array_except_self.py
# 238 : Product of Array Except Self
# https://leetcode.com/problems/product-of-array-except-self/


class Solution:
    # SOLUTION
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        result, suf, pre = [1]*len(nums), 1, 1

        for i in range(len(nums)):
            result[i] *= pre
            pre *= nums[i]
            result[-1-i] *= suf
            suf *= nums[-1-i]
        
        return result


if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [1,2,3,4]

    # OUTPUT
    result = o.productExceptSelf(nums)
    print(result)